File: MaxHeap Sort/Heap.java
import java.util.*;
class Heap{
	
	public static void main(String args[]){
		ArrayList<Double> A = new ArrayList<Double>();
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		for(int i=0; i<n; A.add(sc.nextDouble()),i++);
		Heap h = new Heap();
		h.sort(A);
		for(int i=0; i<n; System.out.print(A.get(i++) +" "));
	}

	void sort(ArrayList<Double> A){ 
		offlineInsert(A);
		for(int i=A.size()-1; i>=0;i--){
			A.set(i, extractMax());
		}
	}

	/*  left child of i: 2i+1
		right child of i: 2i+2
		parent of i: (i-1)/2
	*/
	ArrayList<Double> heap; 
	Heap(){
		heap=new ArrayList<Double>();
	}

	public void onlineInsert(double x){
		heap.add(x);
		int i= heap.size()-1;
		while(i>0 && heap.get(i) > heap.get((i-1)/2)){
			swap(i,(i-1)/2);
			i = (i-1)/2;
		}
	}

	public void offlineInsert(ArrayList<Double> A){
		for(int i=0; i<A.size(); heap.add(A.get(i++)));
		for(int i=(A.size()-1)/2; i>=0; i--){
			int j = i;
			while(j<heap.size()){
				int newJ = maxHeapify(j);
				if(newJ==j)break;
				else j=newJ;
			}
		}
	}

	public double extractMax(){
		double x = max();
		heap.set(0, heap.get(heap.size()-1));
		heap.remove(heap.size()-1);
		int i=0;
		while(i<heap.size()){
			int newI = maxHeapify(i);
			if(newI==i)
				break;
			else
				i = newI;
		}
		return x;
	}

	private int maxHeapify(int i){
		double p = heap.get(i);
		Double lc = (2*i+1 < heap.size()) ? heap.get(2*i+1) : Double.NEGATIVE_INFINITY;
		Double rc = (2*i+2 < heap.size()) ? heap.get(2*i+2) : Double.NEGATIVE_INFINITY;
		if(lc > p && lc >= rc){
			swap(i, 2*i+1);
			return 2*i+1;
			
		}else if(rc >p && rc >= lc){
			swap(i, 2*i+2);
			return 2*i+2;
		}
		return i;
	}

	public double max(){
		return heap.get(0); 
	}

	public void swap(int i, int j){
		double t = heap.get(i);
		heap.set(i, heap.get(j));
		heap.set(j, t);
	}
}

File: MaxHeap Sort/heap.py
class Heap:
	def __init__(self):
		self.heap=[]

	def onlineInsert(self, x):
		self.heap.append(x)
		i = len(self.heap)-1
		while i>0 and self.heap[i] > self.heap[(i-1)//2]:
			self.heap[i], self.heap[(i-1)//2] = self.heap[(i-1)//2], self.heap[i]
			i = (i-1)//2

	def max(self):
		return self.heap[0]

	def maxHeapify(self, i):
		p = self.heap[i]
		lc = self.heap[2*i+1] if 2*i+1 < len(self.heap) else float("-inf")
		rc = self.heap[2*i+2] if 2*i+2 < len(self.heap) else float("-inf")
		if lc > p and lc >= rc:
			self.heap[i], self.heap[2*i+1] = self.heap[2*i+1], self.heap[i]
			return 2*i+1
		elif rc > p and rc >=lc:
			self.heap[i], self.heap[2*i+2] = self.heap[2*i+2], self.heap[i]
			return 2*i+2
		else:
			return i

	
	def offlineInsert(self, A, mH):
		self.heap.extend(A)
		i = (len(A)-1)//2
		while i>=0:
			j=i
			while i<len(self.heap):
				newJ = mH(j)
				if j==newJ:
					break
				else:
					j=newJ
			i-=1
	
	
	def extractMax(self, max, mH):
		x = max()
		self.heap[0] = self.heap[len(self.heap)-1]
		self.heap.pop()
		i=0
		while i<len(self.heap):
			newI = mH(i)
			if newI==i:
				break
			else:
				i=newI
		return x


def sort(A):
	h = Heap()
	h.offlineInsert(A, h.maxHeapify)
	i=len(A)-1
	while i>=0:
		A[i] = h.extractMax(h.max, h.maxHeapify)
		i-=1

n = int(input())
i=0
A=[]
while(i<n):
	A.append(float(input()))
	i+=1

sort(A)
for x in A:
	print(x),